### PR TITLE
[WIP] GH-89727: Fix `shutil.rmtree()` recursion error on deep trees.

### PR DESCRIPTION
**Pull request by barneygale**
Add a `follow_junctions` argument to `pathlib.Path.walk()`. When set to false, directory junctions are treated as files.

Add an `_fwalk()` method to `pathlib.Path`, similar to `os.fwalk()` except in its treatment of symlinks.

(This method could be made public in future if desired).

Implement `shutil.rmtree()` using `pathlib.Path.walk()` or `_fwalk()`.

- Issue: gh-89727

### CODE REVIEW

File: Lib/shutil.py [hunk 1]
```diff
@@ -665,6 +665,8 @@ def onexc(*args):
     if _use_fd_functions:
         return _rmtree_safe_fd(dir_fd, path, onexc)
     else:
+        if dir_fd is not None:
```
Comment (arhadthedev) [inline Lib/shutil.py:661]: Should it be

```suggestion
        if dir_fd is None:
```

?

Comment (barneygale) [reply Lib/shutil.py:661]: The situation arises if a user on a system that doesn't support `dir_fd` (e.g. Windows) calls `shutil.rmtree('blah', dir_fd=somevalue)`, where *somevalue* isn't none. I think the condition is correct!

Comment (arhadthedev) [reply Lib/shutil.py:661]: My apologises, I have read the chunk out of context (skipping the code above to save time). As a result, I misinterpreted the intention of the chunk as *if `dir_fd` is not available, we cannot continue so report and abort*.

**Comment by barneygale** [comment]
Withdrawing this PR - it's out of date, and I'd like to keep the `Path` API stable until `PathBase` is made public (see #110109)

**Pull request closed without merging**